File: PackingTool/Client/src/models/packing/item/PackingItem.ts
import type { IPackingItem } from "./IPackingItem"
import type { PackingItemAttribute } from "./PackingItemAttribute"

export class PackingItem implements IPackingItem {
  readonly ID: number
  Name: string
  Count: number
  Sort: number
  Attributes: PackingItemAttribute[]
  IsNew: boolean
  Packed: boolean

  private constructor(
    id: number,
    name: string,
    count: number,
    sort: number,
    attributes: PackingItemAttribute[],
    isNew: boolean,
    packed: boolean
  ) {
    this.ID = id
    this.Name = name
    this.Count = count
    this.Sort = sort
    this.Attributes = attributes
    this.IsNew = isNew
    this.Packed = packed
  }

  static New(
    id: number,
    name: string,
    count: number,
    sort: number,
    attributes: PackingItemAttribute[]
  ) {
    return new PackingItem(id, name, count, sort, attributes, true, false)
  }

  static Undefined() {
    return new PackingItem(
      0,
      "",
      1,
      1,
      [] as PackingItemAttribute[],
      true,
      false
    )
  }

  static FromJson(json: IPackingItem) {
    return new PackingItem(
      json.ID,
      json.Name,
      json.Count,
      json.Sort,
      json.Attributes ?? ([] as PackingItemAttribute[]),
      false,
      json.Packed ?? false
    )
  }

  ToJson() {
    return <IPackingItem>{
      ID: this.ID,
      Name: this.Name,
      Count: this.Count,
      Sort: this.Sort,
      Attributes: this.Attributes,
      Packed: this.Packed,
    }
  }

  Synchronize(otherItem: IPackingItem) {
    this.Name = otherItem.Name
    this.Count = otherItem.Count
    this.Sort = otherItem.Sort
    this.Attributes = [...otherItem.Attributes]
    this.Packed = otherItem.Packed
  }
}

File: PackingTool.Core/Service/PackingList/Json/Item/PackingItem.cs
using System.ComponentModel.DataAnnotations;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;

namespace PackingTool.Core.Service.PackingList.Json.Item
{
    public class PackingItem
    {
        [Required]
        public int ID { get; set; }
        [Required]
        public string Name { get; set; }
        [Required]
        public int Count { get; set; }
        [Required]
        public int Sort { get; set; }
        [Required]
        [JsonProperty(ItemConverterType = typeof(StringEnumConverter))]
        public PackingItemAttribute[] Attributes { get; set; } //TODO do not serialize null
        [Required]
        public bool Packed { get; set; }

        public PackingItem()
        {
            ID = 0;
            Name = string.Empty;
            Count = 0;
            Sort = 0;
            Attributes = Array.Empty<PackingItemAttribute>();
            Packed = false;
        }

        [JsonConstructor]
        private PackingItem(
            int id,
            string name,
            int count,
            int sort,
            PackingItemAttribute[] attributes,
            bool packed
        )
        {
            ID = id;
            Name = name;
            Count = count;
            Sort = sort;
            Attributes = attributes ?? Array.Empty<PackingItemAttribute>();
            Packed = packed;
        }

        public bool IsValid()
        {
            if (ID <= 0)
            {
                throw new InvalidDataException(
                    $"{nameof(PackingItem)}: ID is missing."
                );
            }

            if (string.IsNullOrEmpty(Name))
            {
                throw new InvalidDataException(
                    $"{nameof(PackingItem)}: name is missing."
                );
            }

            if (Count < 1)
            {
                throw new InvalidDataException(
                    $"{nameof(PackingItem)}: '{Name}' invalid Count: {Count}."
                );
            }

            if (Attributes == null)
            {
                throw new InvalidDataException(
                    $"{nameof(PackingItem)}: '{Name}' has null Attributes."
                );
            }

            if (Attributes.Contains(PackingItemAttribute.Undefined))
            {
                throw new InvalidDataException(
                    $"{nameof(PackingItem)}: '{Name}' has undefined Attribute."
                );
            }

            return true;
        }
    }
}
